<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Getting started: managing data" />


<title>Module 1.2</title>

<script src="site_libs/header-attrs-2.24/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/journal.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">UNR R workshop, Spring 2024</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Workshop #1: getting started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="module1_1.html">Module 1.1: working with objects</a>
    </li>
    <li>
      <a href="module1_2.html">Module 1.2: managing data</a>
    </li>
    <li>
      <a href="module1_3.html">Module 1.3: plotting in ggplot2</a>
    </li>
    <li>
      <a href="module1_4.html">Module 1.4: data wrangling</a>
    </li>
    <li>
      <a href="EXTRAS.html">Module 1.5: final demo</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Workshop #2: beyond the basics
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="module2_1.html">Module 2.1: R packages</a>
    </li>
    <li>
      <a href="module2_2.html">Module 2.2: Programming in R</a>
    </li>
    <li>
      <a href="module2_3.html">Module 2.3: Stats in R</a>
    </li>
    <li>
      <a href="module2_4.html">Module 2.4: Spatial analysis in R</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Data sets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="data.dat">Example: tab-separated data</a>
    </li>
    <li>
      <a href="data.csv">Example: comma-separated data</a>
    </li>
    <li>
      <a href="data.txt">Example: whitespace-separated data</a>
    </li>
    <li>
      <a href="data_missing.txt">Example: missing data</a>
    </li>
    <li>
      <a href="MTMetStations.csv">Meteorological Data</a>
    </li>
    <li>
      <a href="sculpineggs.csv">Sculpin Data</a>
    </li>
    <li>
      <a href="turtle_data.txt">Turtle Data</a>
    </li>
    <li>
      <a href="geneData.csv">Gene Data</a>
    </li>
    <li>
      <a href="reptiles.csv">Reptile Data</a>
    </li>
    <li>
      <a href="comm_data.txt">Comm Data</a>
    </li>
    <li>
      <a href="data.zip">GIS Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="Links.html">Links</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Module 1.2</h1>
<h4 class="author">Getting started: managing data</h4>
<h4 class="date">Spring 2024</h4>

</div>


<div id="load-script-for-module-1.2" class="section level2">
<h2>Load script for module #1.2</h2>
<ol style="list-style-type: decimal">
<li><p>Click <a href="module1_2.R">here</a> to download the script! Save
the script to the project directory you set up in the previous
module.</p></li>
<li><p>Load your script in RStudio. To do this, open RStudio and click
on the folder icon in the toolbar at the top to load your
script.</p></li>
</ol>
<p>Let’s get started with data management in R!</p>
</div>
<div id="working-directory" class="section level2">
<h2>Working directory</h2>
<p>The working directory is the first place R looks for any files you
would like to read in (e.g., code, data). It is also the first place R
will try to write any files you want to save.</p>
<p>It makes things a lot easier if you put all your data files into a
single folder and tell R to make that folder your <strong>working
directory</strong>. That way, you won’t need to wrangle complex
directory names!</p>
<p>Every R session has a working directory, whether you specify one or
not. Let’s find out what my working directory is right now:</p>
<pre class="r"><code># Working directories --------------------

# Find the directory you&#39;re working in 
getwd()          # note: the results from running this command on my machine will differ from yours!  </code></pre>
<pre><code>## [1] &quot;C:/Users/Kevin/Documents/GitHub/R-Bootcamp&quot;</code></pre>
<p>What’s yours? Usually the default working directory is your
“Documents” folder.</p>
<p>Since you should already be working in an RStudio Project, your
working directory will be set as your project directory (directory that
contains the .Rproj file for your project). This is convenient,
because:</p>
<ol style="list-style-type: decimal">
<li>That’s most likely where the data for that project live anyway<br />
</li>
<li>If you’re collaborating with someone else on the project (e.g., in a
shared Dropbox folder), you can both open the project and R will
instantly know where to read and write data without either of you having
to reset the working directory (even though the directory path is
probably different on your two machines). This can save a lot of
headaches!</li>
</ol>
<p>Reminder: to start a new RStudio Project, just click on “File-&gt;New
Project” in RStudio’s menu bar.</p>
<div id="set-your-working-directory" class="section level3">
<h3>Set your working directory</h3>
<p>You can set a new working directory using the “setwd()” function.</p>
<pre class="r"><code># for example...

# setwd(&quot;E:/GIT/R-Bootcamp&quot;)   # note that the use of backslashes for file paths, as used by Windows, are not supported by R</code></pre>
<p><strong>NOTE:</strong> when you put file paths in R, they need to use
forward slashes (“/”; or double backslashes, “\\”) – single backslashes
(“\”, as seen in Windows) do not work for specifying file paths in
R.</p>
<p>Alternatively, you can (in RStudio) use the dropdown menus at the top
to set the working directory (Session-&gt;Set Working
Directory-&gt;Choose Directory).</p>
<p>Once you have set the working directory, you can use the
“list.files()” function to see what’s in the directory. If I ran this,
we would see the contents of the directory that I’m using to create this
website!</p>
<pre class="r"><code># Contents of working directory
# list.files()    # uncomment this to run it...</code></pre>
<p>What’s in your working directory?</p>
</div>
</div>
<div id="importing-data-into-r" class="section level2">
<h2>Importing data into R!</h2>
<p>There are many ways data can be imported into R. Many types of files
can be imported (e.g., text files, csv, shapefiles). And people are
always inventing new ways to read and write data to/from R. But here are
the basics.</p>
<ul>
<li>read.table or read_delim
<ul>
<li>reads a data file in any major text format (comma-delimited, space
delimited etc.), you can specify which format (very general)<br />
</li>
</ul></li>
<li>read.csv or read_csv
<ul>
<li>fields are separated by a comma (this is the most common way to read
in data)</li>
</ul></li>
</ul>
<p>Before we can read data in, we need to put some data files in our
working directory!</p>
<p>NOTE: you can also read data from Excel (.xlsx files) directly using
the <a
href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-import.pdf">‘readxl’
package</a>.</p>
<ol style="list-style-type: decimal">
<li><p>Download the following data files and store them in your working
directory (i.e., the folder where your scripts are already!)</p>
<ul>
<li><a href="data.txt">Whitespace delimited data file</a><br />
</li>
<li><a href="data.csv">Comma delimited data file</a></li>
</ul></li>
<li><p>Make sure the data files are stored in your working directory
(project directory)!</p></li>
</ol>
<p>Once you have saved these files to your working directory, open one
or two of them up (e.g., in Excel or a text editor) to see what’s
inside.</p>
<p>Now let’s read them into R!!</p>
<pre class="r"><code>#  Import/Export data files into R----------------------

# read_csv to import textfile with columns separated by commas
data.df &lt;- read_csv(&quot;data.csv&quot;)
names(data.df) 

# Remove redundant objects from your workspace
rm(data.txt.df)</code></pre>
<div id="using-rs-built-in-data" class="section level3">
<h3>Using R’s built in data</h3>
<p>R has many useful built-in data sets that come pre-loaded. You can
explore these datasets with the following command:</p>
<pre class="r"><code># Built-in data files  -----------------------
data()   </code></pre>
<p>Let’s read in one of these datasets!</p>
<pre class="r"><code>data(mtcars)   # read built-in data on car road tests performed by Motor Trend

head(mtcars)    # inspect the first few lines

# ?mtcars        # learn more about this built-in data set</code></pre>
<p>Many packages come with built-in datasets as well. For, instance,
ggplot2 comes with the “diamonds” package:</p>
<pre class="r"><code>ggplot2::diamonds   # note the use of the package name followed by two colons- this is a way to make sure you are using a function (or data set or other object) from a particular package... (sometimes several packages have functions with the same name...)</code></pre>
<pre><code>## # A tibble: 53,940 × 10
##    carat cut       color clarity depth table price     x     y     z
##    &lt;dbl&gt; &lt;ord&gt;     &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1  0.23 Ideal     E     SI2      61.5    55   326  3.95  3.98  2.43
##  2  0.21 Premium   E     SI1      59.8    61   326  3.89  3.84  2.31
##  3  0.23 Good      E     VS1      56.9    65   327  4.05  4.07  2.31
##  4  0.29 Premium   I     VS2      62.4    58   334  4.2   4.23  2.63
##  5  0.31 Good      J     SI2      63.3    58   335  4.34  4.35  2.75
##  6  0.24 Very Good J     VVS2     62.8    57   336  3.94  3.96  2.48
##  7  0.24 Very Good I     VVS1     62.3    57   336  3.95  3.98  2.47
##  8  0.26 Very Good H     SI1      61.9    55   337  4.07  4.11  2.53
##  9  0.22 Fair      E     VS2      65.1    61   337  3.87  3.78  2.49
## 10  0.23 Very Good H     VS1      59.4    61   338  4     4.05  2.39
## # ℹ 53,930 more rows</code></pre>
</div>
</div>
<div id="basic-data-checking" class="section level2">
<h2>Basic data checking</h2>
<p>To learn more about the ‘internals’ of any data object in R, we can
use the “str()” (structure) function:</p>
<pre class="r"><code># Check/explore data objects --------------------------

# ?str: displays the internal structure of the data object
str(mtcars)</code></pre>
<pre><code>## &#39;data.frame&#39;:    32 obs. of  11 variables:
##  $ mpg : num  21 21 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 ...
##  $ cyl : num  6 6 4 6 8 6 8 4 4 6 ...
##  $ disp: num  160 160 108 258 360 ...
##  $ hp  : num  110 110 93 110 175 105 245 62 95 123 ...
##  $ drat: num  3.9 3.9 3.85 3.08 3.15 2.76 3.21 3.69 3.92 3.92 ...
##  $ wt  : num  2.62 2.88 2.32 3.21 3.44 ...
##  $ qsec: num  16.5 17 18.6 19.4 17 ...
##  $ vs  : num  0 0 1 1 0 1 0 1 1 1 ...
##  $ am  : num  1 1 1 0 0 0 0 0 0 0 ...
##  $ gear: num  4 4 4 3 3 3 3 4 4 4 ...
##  $ carb: num  4 4 1 1 2 1 4 2 2 4 ...</code></pre>
<pre class="r"><code>str(data.df)</code></pre>
<pre><code>## spc_tbl_ [20 × 4] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
##  $ Country: chr [1:20] &quot;Bolivia&quot; &quot;Brazil&quot; &quot;Chile&quot; &quot;Colombia&quot; ...
##  $ Import : num [1:20] 46 74 89 77 84 89 68 70 60 55 ...
##  $ Export : num [1:20] 0 0 16 16 21 15 14 6 13 9 ...
##  $ Product: chr [1:20] &quot;N&quot; &quot;N&quot; &quot;N&quot; &quot;A&quot; ...
##  - attr(*, &quot;spec&quot;)=
##   .. cols(
##   ..   Country = col_character(),
##   ..   Import = col_double(),
##   ..   Export = col_double(),
##   ..   Product = col_character()
##   .. )
##  - attr(*, &quot;problems&quot;)=&lt;externalptr&gt;</code></pre>
<pre class="r"><code>names(diamonds)</code></pre>
<pre><code>##  [1] &quot;carat&quot;   &quot;cut&quot;     &quot;color&quot;   &quot;clarity&quot; &quot;depth&quot;   &quot;table&quot;   &quot;price&quot;  
##  [8] &quot;x&quot;       &quot;y&quot;       &quot;z&quot;</code></pre>
<p>And we can use the “summary()” function to get a brief summary of the
contents of each column in our data frame:</p>
<pre class="r"><code>summary(data.df)</code></pre>
<pre><code>##    Country              Import         Export        Product         
##  Length:20          Min.   :35.0   Min.   : 0.00   Length:20         
##  Class :character   1st Qu.:66.0   1st Qu.: 3.00   Class :character  
##  Mode  :character   Median :74.0   Median : 8.00   Mode  :character  
##                     Mean   :72.1   Mean   : 9.55                     
##                     3rd Qu.:84.0   3rd Qu.:15.25                     
##                     Max.   :91.0   Max.   :23.00</code></pre>
</div>
<div id="exporting-and-saving-data-in-r" class="section level2">
<h2>Exporting and saving data in R</h2>
<p>Reading in data is one thing, but you will probably also want to
write data to your hard drive as well. There are countless reasons for
this- you might want to use an external program to plot your data, you
might want to archive some simulation results.</p>
<p>Again, there are many ways to write data to a file. Here are the
basics!</p>
<div id="exporting-data-as-table" class="section level3">
<h3>Exporting data as table</h3>
<pre class="r"><code># Exporting data (save to hard drive as data file)

# ?write_csv: writes a CSV file to the working directory

newdf &lt;- data.df[,c(&quot;Country&quot;,&quot;Product&quot;)]
   
write_csv(newdf, file=&quot;data_export.csv&quot;)   # export a subset of the data we just read in.</code></pre>
</div>
<div id="saving-and-loading-r-data-objects-binary"
class="section level3">
<h3>Saving (and loading) R data objects (binary)</h3>
<p>The data loaded in your <strong>R environment</strong> are stored in
your computer’s memory as <strong>binary</strong> representations that
are efficient but not human-readable (this is how computers store and
manage data). But sometimes the lack of human readability isn’t a
problem. For example: what if all we want to do is save the data so that
we can load it back into an R session at a later date? In this case we
never need to look at the data <em>outside</em> of R.</p>
<p>To save data we can use the “saveRDS()” and “save()” functions. To
load data we can use the “readRDS()” and “load()” functions:</p>
<p>NOTE: binary R data files should be stored with the extensions
“.RData”:</p>
<pre class="r"><code>#  Saving and loading

# saveRDS: save a single object from the environment to hard disk
# save: saves one or more objects from the environment to hard disk. Must be read back in with the same name.

a &lt;- 1
b &lt;- data.df$Product

saveRDS(b, &quot;Myobject1.rds&quot;)    # use saveRDS to save individual R objects
save(a,b,file=&quot;Myobjects1.RData&quot;)    # use &#39;save&#39; to save sets of objects
save.image(&quot;Myworkspace.RData&quot;)      # use &#39;save.image&#39; to save your entire workspace

rm(a,b)   # remove these objects from the environment

new_b &lt;- readRDS(&quot;Myobject1.rds&quot;)
load(&quot;Myworkspace.RData&quot;)   # load these objects back in with the same names!</code></pre>
</div>
<div id="clearing-the-environment" class="section level3">
<h3>Clearing the environment</h3>
<p>Sometimes your environment can get cluttered with objects. In these
cases, it can help to clear the environment. You can just use the
‘broom’ icon in your environment window in RStudio to clear your
environment, or use this command:</p>
<pre class="r"><code># Clear the environment

rm(list=ls())   # clear the entire environment. Confirm that your environment is now empty!

data.df &lt;- read_csv(&quot;data.csv&quot;)  # read the data back in!</code></pre>
</div>
</div>
<div id="working-with-data-in-r" class="section level2">
<h2>Working with data in R</h2>
<p>Now let’s start seeing what we can do with data in R. Even without
doing any statistical analyses, R is very a powerful environment for
doing data transformations and performing mathematical operations.</p>
<div id="boolean-operations" class="section level3">
<h3>Boolean operations</h3>
<p>Boolean operations refer to TRUE/FALSE tests. That is, we can ask a
question about the data to a Boolean operator and the operator will
return a TRUE or a FALSE (logical) result.</p>
<p>First, let’s meet the boolean operators.</p>
<p>NOTE: don’t get confused between the equals sign (“=”), which is an
<em>assignment operator</em> (same as “&lt;-”), and the double equals
sign (“==”), which is a Boolean operator:</p>
<pre class="r"><code># Working with data in R ------------------------

# &lt;- assignment operator
# =  alternative assignment operator

a &lt;- 3     # assign the value 3 to the object named &quot;a&quot;
b = 5      # assign the value 5 to the object named &quot;b&quot; 
a == 3     # answer the question: &quot;does the object &quot;a&quot; equal &quot;3&quot;? </code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>a == b</code></pre>
<pre><code>## [1] FALSE</code></pre>
<pre class="r"><code># what happens if you accidentally typed &#39;a = b&#39;?</code></pre>
<pre><code>#  Boolean operations

# Basic operators

# &lt;    less than
# &gt;    greater than
# &lt;=   less than or equal to
# &gt;=   greater than or equal to
# ==   equal to
# !=   not equal to
# %in% belongs to a set

# Combining multiple conditions

# &amp;    must meet both conditions (AND operator)
# |    must meet one of two conditions (OR operator)
</code></pre>
<pre class="r"><code># explore Boolean operators -----------------------


Y &lt;- 4   # first define a couple new objects
Z &lt;- 6

Y == Z  # is Y equal to Z?  (T/F)
Y &lt; Z   # is Y less than Z?

!(Y &lt; Z)  # the exclamation point reverses any boolean object (read &quot;NOT&quot;)

data.df[,2]=74     # (for each element in the second column of data.df) is it equal to 74? 

# OOPS! sets entire second column equal to 74! OOPS WE GOOFED UP!!!

data.df &lt;- read.csv(&quot;data.csv&quot;)  ## correct our mistake in the previous line (revert to the original data)!

# Let&#39;s do it right this time
data.df[,2]==74    # tests each element of column to see whether it is equal to 74

data.df[,2]&lt;74|data.df[,2]==91   # combine two questions using the logical OR operator</code></pre>
<p>This image from Wickham and Grolemund’s <a
href="https://r4ds.had.co.nz/index.html">R for Data Science</a> book can
help conceptualize the combination operators:</p>
<p><img src="transform-logical.png" style="width:50.0%" /></p>
</div>
<div id="data-subsetting-using-boolean-logic" class="section level3">
<h3>Data subsetting using boolean logic</h3>
<p>Boolean operations are great for subsetting data - that is, we can
select only those rows/observations that meet a certain condition (e.g.,
where [some condition] is TRUE).</p>
<pre class="r"><code>#  Subsetting data 

data.df[data.df[,&quot;Import&quot;]&lt;74,]    # select those rows of data for which second column is less than 74</code></pre>
<pre><code>## # A tibble: 9 × 4
##   Country      Import Export Product
##   &lt;chr&gt;         &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
## 1 Bolivia          46      0 N      
## 2 DominicanRep     68     14 A      
## 3 Ecuador          70      6 A      
## 4 ElSalvador       60     13 A      
## 5 Guatemala        55      9 A      
## 6 Haiti            35      3 A      
## 7 Honduras         51      7 A      
## 8 Nicaragua        68      0 A      
## 9 Peru             73      0 N</code></pre>
<pre class="r"><code>#  or alternatively, using tidyverse syntax (and the pipe operator):

data.df %&gt;% 
  filter(Import&lt;74)    # use &quot;filter&quot; verb from &#39;dplyr&#39; package</code></pre>
<pre><code>## # A tibble: 9 × 4
##   Country      Import Export Product
##   &lt;chr&gt;         &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
## 1 Bolivia          46      0 N      
## 2 DominicanRep     68     14 A      
## 3 Ecuador          70      6 A      
## 4 ElSalvador       60     13 A      
## 5 Guatemala        55      9 A      
## 6 Haiti            35      3 A      
## 7 Honduras         51      7 A      
## 8 Nicaragua        68      0 A      
## 9 Peru             73      0 N</code></pre>
<pre class="r"><code>sub.countries&lt;-c(&quot;Chile&quot;,&quot;Colombia&quot;,&quot;Mexico&quot;)    # create vector of character strings

data.df %&gt;% 
  filter(Country %in% sub.countries)   # subset the dataset for only that subset of countries we&#39;re interested in</code></pre>
<pre><code>## # A tibble: 3 × 4
##   Country  Import Export Product
##   &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
## 1 Chile        89     16 N      
## 2 Colombia     77     16 A      
## 3 Mexico       83      4 N</code></pre>
</div>
</div>
<div id="practice-data-processing" class="section level2">
<h2>Practice data processing!</h2>
<p>Let’s switch to a different data file. Download the <a
href="turtle_data.txt">Turtle Data</a> and save to your project
directory.</p>
<p>Let’s use this data set to review the summarizing and subsetting
operations we have learned- and learn a few new things along the
way:</p>
<div id="subsetting-data" class="section level3">
<h3>Subsetting data</h3>
<pre class="r"><code># Subsetting data ----------------------

#  Practice subsetting a data frame 
turtles.df &lt;- read_delim(file=&quot;turtle_data.txt&quot;,delim=&quot;\t&quot;)   # tab-delimited file
turtles.df</code></pre>
<pre><code>## # A tibble: 21 × 5
##    tag_number sex    carapace_length head_width weight
##         &lt;dbl&gt; &lt;chr&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1         10 male              41         7.15   7.6 
##  2         11 female            46.4       8.18  11   
##  3          2 &lt;NA&gt;              24.3       4.42   1.65
##  4         15 &lt;NA&gt;              28.7       4.89   2.18
##  5         16 &lt;NA&gt;              32         5.37   3   
##  6          3 female            42.8       7.32   8.6 
##  7          4 male              40         6.6    6.5 
##  8          5 female            45         8.05  10.9 
##  9         12 female            44         7.55   8.9 
## 10         13 &lt;NA&gt;              28         4.85   1.97
## # ℹ 11 more rows</code></pre>
<pre class="r"><code># Subset for turtles that weigh greater than or equal to 10g

subset.turtles.df &lt;- turtles.df %&gt;% 
  filter(weight &gt;= 10)

subset.turtles.df</code></pre>
<pre><code>## # A tibble: 4 × 5
##   tag_number sex    carapace_length head_width weight
##        &lt;dbl&gt; &lt;chr&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
## 1         11 female            46.4       8.18   11  
## 2          5 female            45         8.05   10.9
## 3         22 female            48.1       8.55   12.8
## 4          7 female            48         8.67   13.5</code></pre>
<pre class="r"><code># Subset for only females

fem.turtles.df = turtles.df %&gt;%
  filter(sex==&quot;female&quot;)
  
# Here we want to know the mean weight of all females 
mean(fem.turtles.df$weight)</code></pre>
<pre><code>## [1] 10.02857</code></pre>
<pre class="r"><code># or we can summarize the mean weight for males and females at the same time using the following tidyverse syntax ...

turtles.df %&gt;% 
  group_by(sex) %&gt;% 
  summarize(meanwt = mean(weight))</code></pre>
<pre><code>## # A tibble: 4 × 2
##   sex    meanwt
##   &lt;chr&gt;   &lt;dbl&gt;
## 1 fem      6.2 
## 2 female  10.0 
## 3 male     7.29
## 4 &lt;NA&gt;     2.35</code></pre>
<pre class="r"><code># OOPS, looks like we just caught an error!</code></pre>
</div>
<div id="fixing-data-using-subsetting" class="section level3">
<h3>Fixing data using subsetting</h3>
<p>Let’s fix the error we just noticed in the way sex was represented in
the turtle dataset.</p>
<pre class="r"><code>unique(turtles.df$sex)  # note the two ways of representing females...</code></pre>
<pre><code>## [1] &quot;male&quot;   &quot;female&quot; NA       &quot;fem&quot;</code></pre>
<pre class="r"><code>turtles.df$sex[turtles.df$sex==&quot;fem&quot;] &lt;- &quot;female&quot;  # correct the error

# or alternatively
turtles.df = turtles.df %&gt;% 
  mutate(
    sex= fct_collapse(sex,
             female=c(&quot;fem&quot;,&quot;female&quot;),
             male=c(&quot;male&quot;)))
  

# or alternatively
turtles.df = turtles.df %&gt;% 
  mutate(sex = replace(sex,sex==&quot;fem&quot;,&quot;female&quot;))

turtles.df %&gt;%          # summarize weight by sex (check that it&#39;s fixed)
  group_by(sex) %&gt;% 
  summarize(meanwt = mean(weight))</code></pre>
<pre><code>## # A tibble: 3 × 2
##   sex    meanwt
##   &lt;fct&gt;   &lt;dbl&gt;
## 1 female   9.55
## 2 male     7.29
## 3 &lt;NA&gt;     2.35</code></pre>
</div>
<div id="selecting-and-renaming-columns" class="section level3">
<h3>Selecting and renaming columns</h3>
<p>Sometimes we only want to work with a small subset of columns from a
larger dataset. Or we have column names that are not informative or
don’t make sense to us.</p>
<pre class="r"><code># select only the x y and z columns from the diamonds dataset

diamonds2 &lt;- diamonds %&gt;% 
  select(x,y,z)
 
# in base R (non tidyverse) you can do this:
diamonds2 &lt;- diamonds[,c(&quot;x&quot;,&quot;y&quot;,&quot;z&quot;)]

# to change the column names, use the &quot;names()&quot; function

names(diamonds2)  # extract the column names</code></pre>
<pre><code>## [1] &quot;x&quot; &quot;y&quot; &quot;z&quot;</code></pre>
<pre class="r"><code>names(diamonds2)  &lt;- c(&quot;length&quot;, &quot;width&quot;,&quot;depth&quot;)

# or rename specific variables using &#39;dplyr&#39; from the &#39;tidyverse;

diamonds2 &lt;- rename(diamonds2, len = length)</code></pre>
</div>
<div id="sortingordering-data" class="section level3">
<h3>Sorting/ordering data</h3>
<p><strong>sorting</strong> is another common data operation, which
helps to visualize and organize data. In R, sorting is typically
accomplished using the “order()” function:</p>
<p>*order()** returns the indices of the original (unsorted) vector in
the order that they would appear if properly sorted (increasing, by
default) – i.e., “10,3,22” becomes “2,1,3” (i.e., to sort this vector in
increasing order, you would need to take the second element, then the
first, and then the third).</p>
<p>In the ‘tidyverse’ we can use the “arrange” verb to do this!</p>
<pre class="r"><code>#  Sorting  

# The &#39;order&#39; function returns the indices of the original (unsorted) vector in the order that would sort the vector from lowest to highest...   
order(turtles.df$carapace_length)</code></pre>
<pre><code>##  [1]  3 10  4 20  5 12 13 14  7 11  1 15  6 18  9 17 21  8  2 19 16</code></pre>
<pre class="r"><code># To sort a data frame by one vector, you can use &quot;order()&quot;
turtles.df[order(turtles.df$tag_number),]</code></pre>
<pre><code>## # A tibble: 21 × 5
##    tag_number sex    carapace_length head_width weight
##         &lt;dbl&gt; &lt;fct&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1          1 &lt;NA&gt;              29.2       5.1    2.38
##  2          2 &lt;NA&gt;              24.3       4.42   1.65
##  3          3 female            42.8       7.32   8.6 
##  4          4 male              40         6.6    6.5 
##  5          5 female            45         8.05  10.9 
##  6          6 female            40         6.53   6.2 
##  7          7 female            48         8.67  13.5 
##  8          8 &lt;NA&gt;              32         5.35   2.9 
##  9          9 male              35         5.74   3.9 
## 10         10 male              41         7.15   7.6 
## # ℹ 11 more rows</code></pre>
<pre class="r"><code># And in decreasing order:

turtles.df[order(turtles.df$tag_number,decreasing=T),]</code></pre>
<pre><code>## # A tibble: 21 × 5
##    tag_number sex    carapace_length head_width weight
##         &lt;dbl&gt; &lt;fct&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1        105 male              44         7.1    9   
##  2        104 male              44         7.35   9   
##  3         22 female            48.1       8.55  12.8 
##  4         19 male              42.3       6.77   7.8 
##  5         17 female            35.1       6.04   4.5 
##  6         16 &lt;NA&gt;              32         5.37   3   
##  7         15 &lt;NA&gt;              28.7       4.89   2.18
##  8         14 male              43         6.6    7.2 
##  9         13 &lt;NA&gt;              28         4.85   1.97
## 10         12 female            44         7.55   8.9 
## # ℹ 11 more rows</code></pre>
<pre class="r"><code># or we can use the &quot;arrange&quot; verb in the &#39;tidyverse&#39;:
turtles.df %&gt;% 
  arrange(carapace_length)</code></pre>
<pre><code>## # A tibble: 21 × 5
##    tag_number sex    carapace_length head_width weight
##         &lt;dbl&gt; &lt;fct&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1          2 &lt;NA&gt;              24.3       4.42   1.65
##  2         13 &lt;NA&gt;              28         4.85   1.97
##  3         15 &lt;NA&gt;              28.7       4.89   2.18
##  4          1 &lt;NA&gt;              29.2       5.1    2.38
##  5         16 &lt;NA&gt;              32         5.37   3   
##  6          8 &lt;NA&gt;              32         5.35   2.9 
##  7          9 male              35         5.74   3.9 
##  8         17 female            35.1       6.04   4.5 
##  9          4 male              40         6.6    6.5 
## 10          6 female            40         6.53   6.2 
## # ℹ 11 more rows</code></pre>
<pre class="r"><code># or if we want in descending order...

turtles.df %&gt;% 
  arrange(desc(carapace_length))</code></pre>
<pre><code>## # A tibble: 21 × 5
##    tag_number sex    carapace_length head_width weight
##         &lt;dbl&gt; &lt;fct&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1         22 female            48.1       8.55   12.8
##  2          7 female            48         8.67   13.5
##  3         11 female            46.4       8.18   11  
##  4          5 female            45         8.05   10.9
##  5         12 female            44         7.55    8.9
##  6        105 male              44         7.1     9  
##  7        104 male              44         7.35    9  
##  8         14 male              43         6.6     7.2
##  9          3 female            42.8       7.32    8.6
## 10         19 male              42.3       6.77    7.8
## # ℹ 11 more rows</code></pre>
<pre class="r"><code># Sorting by 2 columns
turtles.df %&gt;% 
  arrange(sex,weight)</code></pre>
<pre><code>## # A tibble: 21 × 5
##    tag_number sex    carapace_length head_width weight
##         &lt;dbl&gt; &lt;fct&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1         17 female            35.1       6.04    4.5
##  2          6 female            40         6.53    6.2
##  3          3 female            42.8       7.32    8.6
##  4         12 female            44         7.55    8.9
##  5          5 female            45         8.05   10.9
##  6         11 female            46.4       8.18   11  
##  7         22 female            48.1       8.55   12.8
##  8          7 female            48         8.67   13.5
##  9          9 male              35         5.74    3.9
## 10          4 male              40         6.6     6.5
## # ℹ 11 more rows</code></pre>
</div>
</div>
<div id="challenge-exercises" class="section level2">
<h2>Challenge exercises</h2>
<ol style="list-style-type: decimal">
<li><p>Save the <a href="comm_data.txt">“comm_data.txt”</a> file to your
working directory. Read this file in as a data frame. Select only the
following columns: Hab_class, C_DWN, C_UPS (discard the remaining
columns). Finally, rename these columns as: “Class”,“Downstream”, and
“Upstream” respectively. [hint 1: use read_table to read in the file as
a data frame] [hint 2: use the “select” verb to select the columns you
want] [hint 3: use the names() function to rename the columns]</p></li>
<li><p>Read in the file <a href="turtle_data.txt">“turtle_data.txt”</a>.
Create a new version of this data frame with all missing data removed
(discard all rows with one or more missing data). Save this new data
frame to your project directory as a comma delimited text file. [hint 1:
use the “na.omit()” function to remove rows with NAs] [hint 2: use the
write_csv function to write to your working directory]</p></li>
<li><p>Read in the file <a href="turtle_data.txt">“turtle_data.txt”</a>.
Create a new data frame with only male turtles. Use this subsetted data
set to compute the mean and standard deviation for carapace length of
male turtles.</p></li>
</ol>
<pre class="r"><code># CHALLENGE EXERCISES   -------------------------------------

# 1: Save the &quot;comm_data.txt&quot; file to your working directory. Read this file in as a data frame. Select only only the following columns: Hab_class, C_DWN, C_UPS (discard the remaining columns). Finally, rename these columns as: &quot;Class&quot;,&quot;Downstream&quot;, and &quot;Upstream&quot; respectively. [hint 1: use read_table to read in the file as a data frame] [hint 2: use the &quot;select&quot; verb to select the columns you want] [hint 3: use the names() function to rename the columns]
# 
# 2: Read in the file &quot;turtle_data.txt&quot;. Create a new version of this data frame with all missing data removed (discard all rows with one or more missing data). Save this new data frame to your project directory as a comma delimited text file. [hint 1: use the &quot;na.omit()&quot; function to remove rows with NAs] [hint 2: use the write_csv function to write to your working directory] 
#   
# 3: Read in the file &quot;turtle_data.txt&quot;. Create a new data frame with only male turtles. Use this subsetted data set to compute the mean and standard deviation for carapace length of male turtles.</code></pre>
<p><a href="module1_3.html">–go to next submodule–</a></p>
<div id="bonus-dealing-with-missing-data" class="section level3">
<h3>Bonus: dealing with missing data</h3>
<p>In many real-world datasets, observations are incomplete in some way-
they are missing information.</p>
<p>In R, the code “NA” (not available) stands in for elements of a
vector that are missing for whatever reason.</p>
<p>Most statistical functions have ways of dealing with NAs.</p>
<p>Let’s explore a data set with missing data.</p>
<p>Download this <a href="data_missing.txt">Data with missing values</a>
and save to your working directory.</p>
<p>Now let’s explore this data set in more detail:</p>
<pre class="r"><code>#  Missing Data 

# NOTE: you need to specify that this is a tab-delimited file. It is especially important to specify the delimiter for data files with missing data. If you specify the header and what the text is delimited by correctly, it will read missing data as NA. Otherwise it will fail to read data in properly.

missing.df &lt;- read_delim(file=&quot;data_missing.txt&quot;,delim=&quot;\t&quot;)   # try replacing with &quot;read_table&quot;- it does not work right!

# Missing data are read as an NA
missing.df</code></pre>
<pre><code>## # A tibble: 20 × 4
##    Country        Import Export Product
##    &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
##  1 Bolivia            46      0 N      
##  2 Brazil             74      0 N      
##  3 Chile              89     16 N      
##  4 Colombia           77     16 A      
##  5 CostaRica          84     21 A      
##  6 Cuba               89     15 A      
##  7 DominicanRep       NA     14 A      
##  8 Ecuador            70      6 A      
##  9 ElSalvador         60     13 A      
## 10 Guatemala          55      9 A      
## 11 Haiti              35      3 A      
## 12 Honduras           51     NA A      
## 13 Jamaica            87     23 A      
## 14 Mexico             83      4 N      
## 15 Nicaragua          68      0 A      
## 16 Panama             NA     19 N      
## 17 Paraguay           74      3 A      
## 18 Peru               73     NA N      
## 19 TrinidadTobago     84     15 A      
## 20 Venezuela          NA      7 N</code></pre>
<pre class="r"><code># Can summarize your data and tell you how many NA&#39;s per col
summary(missing.df)</code></pre>
<pre><code>##    Country              Import          Export        Product         
##  Length:20          Min.   :35.00   Min.   : 0.00   Length:20         
##  Class :character   1st Qu.:60.00   1st Qu.: 3.25   Class :character  
##  Mode  :character   Median :74.00   Median :11.00   Mode  :character  
##                     Mean   :70.53   Mean   :10.22                     
##                     3rd Qu.:84.00   3rd Qu.:15.75                     
##                     Max.   :89.00   Max.   :23.00                     
##                     NA&#39;s   :3       NA&#39;s   :2</code></pre>
<pre class="r"><code># Omits (removes) rows with missing data
na.omit(missing.df)</code></pre>
<pre><code>## # A tibble: 15 × 4
##    Country        Import Export Product
##    &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
##  1 Bolivia            46      0 N      
##  2 Brazil             74      0 N      
##  3 Chile              89     16 N      
##  4 Colombia           77     16 A      
##  5 CostaRica          84     21 A      
##  6 Cuba               89     15 A      
##  7 Ecuador            70      6 A      
##  8 ElSalvador         60     13 A      
##  9 Guatemala          55      9 A      
## 10 Haiti              35      3 A      
## 11 Jamaica            87     23 A      
## 12 Mexico             83      4 N      
## 13 Nicaragua          68      0 A      
## 14 Paraguay           74      3 A      
## 15 TrinidadTobago     84     15 A</code></pre>
<pre class="r"><code># ?is.na   (Boolean test!)
is.na(missing.df)</code></pre>
<pre><code>##       Country Import Export Product
##  [1,]   FALSE  FALSE  FALSE   FALSE
##  [2,]   FALSE  FALSE  FALSE   FALSE
##  [3,]   FALSE  FALSE  FALSE   FALSE
##  [4,]   FALSE  FALSE  FALSE   FALSE
##  [5,]   FALSE  FALSE  FALSE   FALSE
##  [6,]   FALSE  FALSE  FALSE   FALSE
##  [7,]   FALSE   TRUE  FALSE   FALSE
##  [8,]   FALSE  FALSE  FALSE   FALSE
##  [9,]   FALSE  FALSE  FALSE   FALSE
## [10,]   FALSE  FALSE  FALSE   FALSE
## [11,]   FALSE  FALSE  FALSE   FALSE
## [12,]   FALSE  FALSE   TRUE   FALSE
## [13,]   FALSE  FALSE  FALSE   FALSE
## [14,]   FALSE  FALSE  FALSE   FALSE
## [15,]   FALSE  FALSE  FALSE   FALSE
## [16,]   FALSE   TRUE  FALSE   FALSE
## [17,]   FALSE  FALSE  FALSE   FALSE
## [18,]   FALSE  FALSE   TRUE   FALSE
## [19,]   FALSE  FALSE  FALSE   FALSE
## [20,]   FALSE   TRUE  FALSE   FALSE</code></pre>
<pre class="r"><code>complete.cases(missing.df)   # Boolean: for each row, tests if there are no NA values</code></pre>
<pre><code>##  [1]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE FALSE
## [13]  TRUE  TRUE  TRUE FALSE  TRUE FALSE  TRUE FALSE</code></pre>
<pre class="r"><code># Replace all missing values in the data frame with a an interpolation function from tidyverse: replace_na

missing.df %&gt;% 
  mutate(Export = replace_na(Export,mean(Export,na.rm=T)),
         Import = replace_na(Import,mean(Import,na.rm=T)))</code></pre>
<pre><code>## # A tibble: 20 × 4
##    Country        Import Export Product
##    &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
##  1 Bolivia          46      0   N      
##  2 Brazil           74      0   N      
##  3 Chile            89     16   N      
##  4 Colombia         77     16   A      
##  5 CostaRica        84     21   A      
##  6 Cuba             89     15   A      
##  7 DominicanRep     70.5   14   A      
##  8 Ecuador          70      6   A      
##  9 ElSalvador       60     13   A      
## 10 Guatemala        55      9   A      
## 11 Haiti            35      3   A      
## 12 Honduras         51     10.2 A      
## 13 Jamaica          87     23   A      
## 14 Mexico           83      4   N      
## 15 Nicaragua        68      0   A      
## 16 Panama           70.5   19   N      
## 17 Paraguay         74      3   A      
## 18 Peru             73     10.2 N      
## 19 TrinidadTobago   84     15   A      
## 20 Venezuela        70.5    7   N</code></pre>
<pre class="r"><code># or using tidyverse trickery (less code repetition)
missing.df %&gt;% 
  mutate(across(where(is.numeric),~replace_na(.,mean(.,na.rm=T))))</code></pre>
<pre><code>## # A tibble: 20 × 4
##    Country        Import Export Product
##    &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
##  1 Bolivia          46      0   N      
##  2 Brazil           74      0   N      
##  3 Chile            89     16   N      
##  4 Colombia         77     16   A      
##  5 CostaRica        84     21   A      
##  6 Cuba             89     15   A      
##  7 DominicanRep     70.5   14   A      
##  8 Ecuador          70      6   A      
##  9 ElSalvador       60     13   A      
## 10 Guatemala        55      9   A      
## 11 Haiti            35      3   A      
## 12 Honduras         51     10.2 A      
## 13 Jamaica          87     23   A      
## 14 Mexico           83      4   N      
## 15 Nicaragua        68      0   A      
## 16 Panama           70.5   19   N      
## 17 Paraguay         74      3   A      
## 18 Peru             73     10.2 N      
## 19 TrinidadTobago   84     15   A      
## 20 Venezuela        70.5    7   N</code></pre>
</div>
<div id="aside-the-pipe-operator" class="section level3">
<h3>Aside: the pipe operator (%&gt;%)</h3>
<p>The ‘tidyverse’ set of packages takes advantage of the pipe operator
<code>%&gt;%</code>, which provides a clean and intuitive way to
structure code and perform sequential operations in R.</p>
<p><img src="magritte.jpg" style="width:25.0%" /></p>
<p>Key advantages include:</p>
<ul>
<li><p>code is more readable and intuitive – reads left to right, rather
than inside out as is the case for nested function</p></li>
<li><p>perform multiple operations without creating a bunch of
intermediate (temporary) datasets</p></li>
</ul>
<p>This operator comes from the <em>magrittr</em> package, which is
included in the installation of all of the tidyverse packages. The
shortcut for the pipe operator is ctrl-shift-m (‘m’ is for Magritte).
When reading code out loud, use ‘then’ for the pipe. For example the
command here:</p>
<pre><code>
x %&gt;% log() %&gt;% round(digits=2)
</code></pre>
<p>can be interpreted as follows:</p>
<blockquote>
<p>Take “x”, THEN take its natural logarithm, THEN round the resulting
value to 2 decimal places</p>
</blockquote>
<p>The structure is simple. Start with the object you want to
manipulate, and apply actions (e.g., functions) to that object in the
order in which you want to apply them.</p>
<p>Here is a quick example.</p>
<pre class="r"><code>#  ASIDE: using the pipe operator %&gt;% (ctrl-shift-m) in R 

# start with a simple example
x &lt;- 3

# calculate the log of x
log(x) # form f(x) is equivalent to</code></pre>
<pre><code>## [1] 1.098612</code></pre>
<pre class="r"><code>x %&gt;% log() # form x %&gt;% f</code></pre>
<pre><code>## [1] 1.098612</code></pre>
<pre class="r"><code># example of multiple steps in pipe
round(log(x), digits=2) # form g(f(x)) is equivalent to</code></pre>
<pre><code>## [1] 1.1</code></pre>
<pre class="r"><code>x %&gt;% log() %&gt;% round(digits=2) </code></pre>
<pre><code>## [1] 1.1</code></pre>
</div>
<div id="some-more-subsetting-practice" class="section level3">
<h3>Some more subsetting practice:</h3>
<p>Returning to the turtle example, we can use subsetting operations to
correct or alter data:</p>
<pre class="r"><code>#  Data Manipulation using subsetting 

# list of tags we do not trust the data for
bad.tags &lt;- c(13,105)

turtles.df = turtles.df %&gt;% 
  mutate(
    sex = replace(sex,tag_number%in%bad.tags,NA),
    carapace_length = replace(carapace_length,tag_number%in%bad.tags,NA),
    head_width = replace(head_width,tag_number%in%bad.tags,NA),
    weight = replace(weight,tag_number%in%bad.tags,NA)
  )

# or... use some more tidyverse helper functions and tricks!

turtles.df = turtles.df %&gt;% 
  mutate(across(c(&quot;sex&quot;,&quot;carapace_length&quot;,&quot;head_width&quot;,&quot;weight&quot;),
                ~replace(.,tag_number%in%bad.tags,NA)))

# or use the following non-tidyverse syntax... which still seems easier to me!

turtles.df[turtles.df$tag_number%in%bad.tags,c(&quot;sex&quot;,&quot;carapace_length&quot;,&quot;head_width&quot;,&quot;weight&quot;)]  &lt;- NA

turtles.df</code></pre>
<pre><code>## # A tibble: 21 × 5
##    tag_number sex    carapace_length head_width weight
##         &lt;dbl&gt; &lt;fct&gt;            &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1         10 male              41         7.15   7.6 
##  2         11 female            46.4       8.18  11   
##  3          2 &lt;NA&gt;              24.3       4.42   1.65
##  4         15 &lt;NA&gt;              28.7       4.89   2.18
##  5         16 &lt;NA&gt;              32         5.37   3   
##  6          3 female            42.8       7.32   8.6 
##  7          4 male              40         6.6    6.5 
##  8          5 female            45         8.05  10.9 
##  9         12 female            44         7.55   8.9 
## 10         13 &lt;NA&gt;              NA        NA     NA   
## # ℹ 11 more rows</code></pre>
<pre class="r"><code> # make a new variable &quot;size.class&quot; based on the &quot;weight&quot; variable

turtles.df = turtles.df %&gt;% 
  mutate(size.class = case_when(
    weight &lt; 3 ~ &quot;juvenile&quot;,
    weight &gt; 6 ~ &quot;adult&quot;,
    is.na(weight) ~ NA_character_,
    .default = &quot;subadult&quot;
  ))

turtles.df$size.class</code></pre>
<pre><code>##  [1] &quot;adult&quot;    &quot;adult&quot;    &quot;juvenile&quot; &quot;juvenile&quot; &quot;subadult&quot; &quot;adult&quot;   
##  [7] &quot;adult&quot;    &quot;adult&quot;    &quot;adult&quot;    NA         &quot;adult&quot;    &quot;juvenile&quot;
## [13] &quot;subadult&quot; &quot;subadult&quot; &quot;adult&quot;    &quot;adult&quot;    NA         &quot;adult&quot;   
## [19] &quot;adult&quot;    &quot;juvenile&quot; &quot;adult&quot;</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
